crates/bin/pd/src/migrate/testnet78.rs

```diff
@@ -0,0 +1,392 @@
+                penumbra_governance::proposal_state::Outcome::Passed => {}
+                penumbra_governance::proposal_state::Outcome::Failed { withdrawn } => {
+                    match withdrawn {
+                        penumbra_governance::proposal_state::Withdrawn::No => {}
+                        penumbra_governance::proposal_state::Withdrawn::WithReason { reason } => {
+                            proposal_state = penumbra_governance::proposal_state::State::Claimed {
+                                outcome: penumbra_governance::proposal_state::Outcome::Failed {
+                                    withdrawn:
+                                        penumbra_governance::proposal_state::Withdrawn::WithReason {
+                                            reason: truncate(&reason, 1024).to_string(),
+                                        },
+                                },
+                            };
+                        }
+                    }
+                }
+                penumbra_governance::proposal_state::Outcome::Slashed { withdrawn } => {
+                    match withdrawn {
+                        penumbra_governance::proposal_state::Withdrawn::No => {}
+                        penumbra_governance::proposal_state::Withdrawn::WithReason { reason } => {
+                            proposal_state = penumbra_governance::proposal_state::State::Claimed {
+                                outcome: penumbra_governance::proposal_state::Outcome::Slashed {
+                                    withdrawn:
+                                        penumbra_governance::proposal_state::Withdrawn::WithReason {
+                                            reason: truncate(&reason, 1024).to_string(),
+                                        },
+                                },
+                            };
+                        }
+                    }
+                }
+            },
+        }
+
+        // Store the truncated proposal state data
+        delta.put(
+            penumbra_governance::state_key::proposal_state(proposal_id),
+            proposal_state.clone(),
+        );
+    }
+    Ok(())
+}
+
+// Since the limits are based on `String::len`, which returns
+// the number of bytes, we need to truncate the UTF-8 strings at the
+// correct byte boundaries.
+//
+// This can be simplified once https://github.com/rust-lang/rust/issues/93743
+// is stabilized.
+#[inline]
+pub fn floor_char_boundary(s: &str, index: usize) -> usize {
+    if index >= s.len() {
+        s.len()
+    } else {
+        let lower_bound = index.saturating_sub(3);
+        let new_index = s.as_bytes()[lower_bound..=index]
+            .iter()
+            .rposition(|b| is_utf8_char_boundary(*b));
+
+        // SAFETY: we know that the character boundary will be within four bytes
+        unsafe { lower_bound + new_index.unwrap_unchecked() }
```

Comment:
```suggestion
        lower_bound + new_index.unwrap()
```

Why don't we just avoid the unsafe thing altogether? Seems like we'd prefer a panic in migration testing, if for some reason this code is mistaken (this does seem safe to me, but I don't see the benefit in using an unchecked method)